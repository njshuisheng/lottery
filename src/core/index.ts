export * from './red'
export * from './blue'
export * from './common'