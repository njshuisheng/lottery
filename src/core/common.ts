import { Lottery, Red } from "types"
import { redPrimeComposite } from "./red"

export const num = (v: string | number) => isNaN(Number(v)) ? 0 : Number(v)

export const isPrime = (v: number): boolean => {
    for (let i = 2; i < v; i++) {
        if (v % i === 0) return false
    }

    return true
}

export const getLotteries = (size: number = 10) => {
    return new Promise<Lottery[]>((resolve, reject) => {
        fetch('../../mock/lotteries.json').then(async res => {
            const data: { RECORDS: Lottery[] } = await res.json()
            resolve(data.RECORDS.slice(0, size))
        }).catch(err => reject(err))
    })
}

const data = await getLotteries(100)
const reds: Red[][] = []
data.forEach(n => reds.push(JSON.parse(n.reds)))
const { prime, composite } = redPrimeComposite(reds, 1)
console.log(prime, composite)