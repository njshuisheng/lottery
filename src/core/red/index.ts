import type { Red } from 'types'
import { isPrime, num } from '../common'

export function redSum(reds: Red[]): number {
    return reds.reduce((sum, n) => (sum += num(n), sum), 0)
}

export function redDistance(reds: Red[]): number {
    return Math.max(...reds.map(n => num(n))) - Math.min(...reds.map(n => num(n)))
}

export function redAreaRatio(reds: Red[]) {
    const one = [], two = [], three = []

    for (let i = 0; i < reds.length; i++) {
        const red = num(reds[i])
        red <= 11 ? one.push(red) : red <= 22 ? two.push(red) : three.push(red)
    }

    return `${one.length}:${two.length}:${three.length}`
}

// 质合比
export function redPrimeComposite(reds: Red[][], index: number) {
    const prime = [], composite = []
    index = index - 1
    for (let i = 0; i < reds.length; i++) {
        const red = num(reds[i][index])
        isPrime(red) ? prime.push(red) : composite.push(red)
    }

    return { prime: (prime.length / reds.length).toFixed(2), composite: (composite.length / reds.length).toFixed(2) }
}