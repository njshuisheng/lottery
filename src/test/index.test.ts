import { isPrime } from '../core/common'
import { assert, describe, expect, it } from 'vitest'
import { redAreaRatio, redSum, getLotteries } from '../core'

describe('red test', () => {
  it('foo', () => {
    expect(redSum(['01', '02', '04'])).toBe(7)
    expect(true).to.be.true
  })

  it('redAreaRatio', () => {
    expect(redAreaRatio(['01', '02', '16', '23', '28', '33'])).toBe('2:1:3')
  })

  it('prime', async () => {
    // assert.equal(isPrime(5), true)
    // const data = await getLotteries(10)
    // console.log(data)
  })

  it('snapshot', () => {
    expect({ foo: 'bar' }).toMatchSnapshot()
  })
})