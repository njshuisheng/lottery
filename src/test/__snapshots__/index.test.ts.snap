// Vitest Snapshot v1

exports[`red test > snapshot 1`] = `
{
  "foo": "bar",
}
`;

exports[`suite name > snapshot 1`] = `
{
  "foo": "bar",
}
`;
